pkg/controller/acicontainersoperator.go

```diff
@@ -684,3 +827,235 @@ func (c *Controller) handleDaemonsetDelete(obj interface{}) bool {
+			continue
+		}
+		address := node.Status.Addresses
+		for _, val := range address {
+			if val.Type == v1.NodeInternalIP {
+				nodeAddress = append(nodeAddress, val.Address)
+			}
+		}
+	}
+	return nodeAddress, nil
+}
+
+func (c *Controller) getDnsInfo() (*operatorv1.DNS, error) {
+	dnsInfo := &operatorv1.DNS{
+		TypeMeta:   metav1.TypeMeta{APIVersion: operatorv1.GroupVersion.String(), Kind: "DNS"},
+		ObjectMeta: metav1.ObjectMeta{Name: "default"},
+	}
+	err := c.DnsOperatorClient.Get(context.TODO(), types.NamespacedName{
+		Name: "default"}, dnsInfo)
+	if err != nil {
+		log.Info(err)
+		return nil, err
+	}
+	return dnsInfo, nil
+}
+
+// it reads all the node ip address.
+// updates if there is any changes in the address computed
+func (c *Controller) updateDnsOperatorSpec() bool {
+	if c.DnsOperatorClient == nil {
+		return false
+	}
+	dnsInfo, err := c.getDnsInfo()
+	if err != nil {
+		return true
+	}
+	if len(dnsInfo.Spec.Servers) == 0 {
+		return false
+	}
+	var nodeAddress []string
+	nodeAddress, err = c.getNodeAddress()
+	if err != nil {
+		log.Info(err)
+		return true
+	}
+	if !reflect.DeepEqual(dnsInfo.Spec.Servers[0].ForwardPlugin.Upstreams, nodeAddress) {
+		for _, server := range dnsInfo.Spec.Servers {
+			server.ForwardPlugin.Upstreams = nodeAddress
+		}
+		err = c.DnsOperatorClient.Update(context.TODO(), dnsInfo)
+		if err != nil {
+			log.Info(err)
+			return true
+		}
+	}
+	log.Infof("dnsInfo: %+v", dnsInfo)
+	return false
+}
+
+// handle node create to update the dnsOperatorSpec
+func (c *Controller) handleNodeCreate(obj interface{}) bool {
+	log.Infof("node created")
+	return c.updateDnsOperatorSpec()
+}
+
+// handle node delete
+func (c *Controller) handleNodeDelete(obj interface{}) bool {
+	log.Infof("node Deleted")
+	return c.updateDnsOperatorSpec()
+}
+
+// handle route create
+// local route cache will be updated
+// if route is already present it will ignore silently as it isupdate happend in operator create
+func (c *Controller) handleRouteCreate(obj interface{}) bool {
+	route := obj.(*routesv1.Route)
+	log.Infof("route created: %s", route.ObjectMeta.Name)
+	if c.DnsOperatorClient == nil {
+		return false
+	}
+	key, _ := cache.MetaNamespaceKeyFunc(obj)
+	if _, ok := c.routes[key]; ok {
+		return false
+	}
+	c.indexMutex.Lock()
+	c.routes[key] = true
+	c.indexMutex.Unlock()
+	dnsInfo, err := c.getDnsInfo()
+	if err != nil {
+		return true
+	}
+	var server operatorv1.Server
+	server.Name = key
+	server.Zones = append(server.Zones, route.Spec.Host)
+	// if already computed update the cache
+	if len(dnsInfo.Spec.Servers) > 0 {
+		server.ForwardPlugin.Upstreams = dnsInfo.Spec.Servers[0].ForwardPlugin.Upstreams
+	} else { // compute the node ip's fresh
+		nodeaddr, err := c.getNodeAddress()
+		if err != nil {
+			return true
+		}
+		server.ForwardPlugin.Upstreams = nodeaddr
+	}
+	dnsInfo.Spec.Servers = append(dnsInfo.Spec.Servers, server)
+	err = c.DnsOperatorClient.Update(context.TODO(), dnsInfo)
```

Comment:
If update is successful it will be helpful to info log that it succeeded.

Reply:
same here